### PR TITLE
Remove wrapper script

### PR DESCRIPTION
**Pull request by dfornika**
Addresses #10.

### CODE REVIEW

**Comment by jakirkham** [comment]
So I think this content needs to be moved to `juliarc.jl`. Should be in `$PREFIX/etc/julia/juliarc.jl`.

**Comment by jakirkham** [comment]
Tried to add this code for the `juliarc` instead of the wrapper script and it seems to work ok.

```julia
# $PREFIX/etc/julia/juliarc.jl

JULIA_PREFIX = abspath(joinpath(Base.source_path(), "..", "..", ".."))

if !("JULIA_PKGDIR" in keys(ENV))
    ENV["JULIA_PKGDIR"] = joinpath(JULIA_PREFIX, "share", "julia", "site")
    Pkg.init()
    Pkg.__init__()
    pop!(Base.LOAD_CACHE_PATH)
end

if !("JULIA_HISTORY" in keys(ENV))
    ENV["JULIA_HISTORY"] = joinpath(JULIA_PREFIX, ".julia_history")
end
```

**Comment by jakirkham** [comment]
Added PR ( https://github.com/dfornika/julia-feedstock/pull/1 ) against your branch with this sort of change.

File: recipe/juliarc.jl [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+if !("JULIA_PKGDIR" in keys(ENV))
+    ENV["JULIA_PKGDIR"] = joinpath(JULIA_PREFIX, "share", "julia", "site")
+    Pkg.init()
+    Pkg.__init__()
+    pop!(Base.LOAD_CACHE_PATH)
+end
+
+if !("JULIA_HISTORY" in keys(ENV))
+    ENV["JULIA_HISTORY"] = joinpath(JULIA_PREFIX, ".julia_history")
+end
```
Comment (jakirkham) [inline recipe/juliarc.jl]: So I know I added this and it just imitates the wrapper script that we are replacing, but I don't think it is a good idea. If there is a multiuser `conda` install, we don't want to squash their history into one file shared by all of them. This is bad for user's tracking their own history and is bad if some of them lack permissions to the `conda` install. Instead we should keep them separated per user. Ideally this is done by allowing this to write in the user's  home directory somewhere. This is actually what we do with IPython as well.

TL;DR We should remove these 3 lines.

File: recipe/build.sh [hunk 2]
```diff
@@ -13,7 +13,8 @@ make -j 4 prefix=${PREFIX} MARCH=core2 sysconfigdir=${PREFIX}/etc NO_GIT=1 \
...
  install
 
-mv "$PREFIX/bin/julia" "$PREFIX/bin/julia_"
-cp "$RECIPE_DIR/julia-wrapper.sh" "$PREFIX/bin/julia"
-chmod +x "$PREFIX/bin/julia"
+# Configure juliarc to use conda environment
+cp -f "${RECIPE_DIR}/juliarc.jl" "${PREFIX}/etc/julia/juliarc.jl"
 
+# Populate initial package directory
+julia -e "Pkg.init(); Pkg.__init__()"
```
Comment (jakirkham) [inline recipe/build.sh]: We may need to drop `Pkg` initialization step after all. Appears `conda-build` strips out `.git` directories, which will break the `METADATA` repo. There are some hacks we could use to workaround this, but I'm not sure they are worth it.

xref: https://github.com/conda/conda-build/issues/2360

Comment (jakirkham) [reply recipe/build.sh]: Added PR ( https://github.com/dfornika/julia-feedstock/pull/2 ) to drop this. Should fix the build issue.

**Comment by jakirkham** [comment]
cc @tkelman

File: recipe/juliarc.jl (hunk 1 above)
Comment (tkelman) [reply recipe/juliarc.jl]: that applies to users' packages too doesn't it?

File: recipe/build.sh [hunk 3]
```diff
@@ -13,7 +13,5 @@ make -j 4 prefix=${PREFIX} MARCH=core2 sysconfigdir=${PREFIX}/etc NO_GIT=1 \
  TAGGED_RELEASE_BANNER="conda-forge-julia release" \
  install
 
-mv "$PREFIX/bin/julia" "$PREFIX/bin/julia_"
-cp "$RECIPE_DIR/julia-wrapper.sh" "$PREFIX/bin/julia"
-chmod +x "$PREFIX/bin/julia"
-
+# Configure juliarc to use conda environment
+cp -f "${RECIPE_DIR}/juliarc.jl" "${PREFIX}/etc/julia/juliarc.jl"
```
Comment (tkelman) [inline recipe/build.sh]: should append, not replace

File: recipe/juliarc.jl [hunk 4]
```diff
@@ -0,0 +1,8 @@
+JULIA_PREFIX = abspath(joinpath(Base.source_path(), "..", "..", ".."))
+
+if !("JULIA_PKGDIR" in keys(ENV))
+    ENV["JULIA_PKGDIR"] = joinpath(JULIA_PREFIX, "share", "julia", "site")
+    Pkg.init()
```
Comment (tkelman) [inline recipe/juliarc.jl]: this probably shouldn't be done on every startup, only if it hasn't been initialized yet

File: recipe/build.sh (hunk 3 above)
Comment (jakirkham) [reply recipe/build.sh]: It looks like an empty file with a comment about how we can add stuff to it. Why not just replace it?

File: recipe/juliarc.jl (hunk 1 above)
Comment (jakirkham) [reply recipe/juliarc.jl]: That is a totally different view of package management from the one `conda` is based in. The point of using something like `conda` is to allow for multiple different independent environments where many different versions of packages and their dependencies live. This allows one to easily switch between different development and production environments trivially on a regular basis. Also users can trivially toss aside an environment or create a new one fresh without concerns of other dependencies leaking in. We can't very easily do any of these things if packages live outside the environment.

Comment (jakirkham) [reply recipe/juliarc.jl]: Should add that one can create `conda` environments anywhere. So if a user is working with a multiuser `conda` installation and does not have permissions to the multiuser environments (or wishes not to mess with them), they still can create an environment in their home directory or anywhere else with packages of their choosing.

File: recipe/juliarc.jl (hunk 4 above)
Comment (dfornika) [reply recipe/juliarc.jl]: What's the preferred way to check if the package repository has been initialized?

Comment (jakirkham) [reply recipe/juliarc.jl]: Was thinking about that. It seems to only take a toll the first time (as it is initializing everything from scratch). What sort of penalties would we be taking on for subsequent calls?

Initially we tried to run this once during the build, but there was a `conda-build` issue that blocked it. ( https://github.com/conda/conda-build/issues/2360 ) In the end, not sure we should be running this at build time anyways as it would go stale and make the package larger.

Could add a [`post-link`]( https://conda.io/docs/user-guide/tasks/build-packages/link-scripts.html ) step to handle this instead. That would run this step once only when `julia` gets installed. Seems like the best of both worlds really.

OTOH we could just never initialize `Pkg` at all. Would make it slow for the user the first time they run something with `Pkg`, but that would be the case anyways, right?

FWIW borrowed this strategy from [SO answer]( https://stackoverflow.com/a/44918155 ). Related to this, is it necessary to call `Pkg.__init__()` or is that already handled by `Pkg.init()`?

File: recipe/build.sh [hunk 5]
```diff
@@ -13,7 +13,5 @@ make -j 4 prefix=${PREFIX} MARCH=core2 sysconfigdir=${PREFIX}/etc NO_GIT=1 \
  TAGGED_RELEASE_BANNER="conda-forge-julia release" \
  install
 
-mv "$PREFIX/bin/julia" "$PREFIX/bin/julia_"
-cp "$RECIPE_DIR/julia-wrapper.sh" "$PREFIX/bin/julia"
-chmod +x "$PREFIX/bin/julia"
-
+# Configure juliarc to use conda environment
+"${PREFIX}/etc/julia/juliarc.jl" << "${RECIPE_DIR}/juliarc.jl"
```
Comment (jakirkham) [inline recipe/build.sh]: I think you want `cat "${RECIPE_DIR}/juliarc.jl" >> "${PREFIX}/etc/julia/juliarc.jl"`.

Comment (dfornika) [reply recipe/build.sh]: Oops. Was trying to avoid a 'useless use of cat award.'

File: recipe/juliarc.jl [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+    Pkg.__init__()
```
Comment (jakirkham) [inline recipe/juliarc.jl]: Let's go ahead and drop this and the line above it (i.e. `Pkg.init()`).

File: recipe/build.sh (hunk 3 above)
Comment (tkelman) [reply recipe/build.sh]: it isn't empty on all platforms

File: recipe/juliarc.jl (hunk 1 above)
Comment (tkelman) [reply recipe/juliarc.jl]: won't that force them to reinstall julia? packages are user data, not part of the language, and shouldn't be forced to be installed in the same place

Comment (tkelman) [reply recipe/juliarc.jl]: conda doesn't manage julia packages - you can do what you describe in julia by setting different values for JULIA_PKGDIR.

File: recipe/juliarc.jl [hunk 7]
```diff
@@ -0,0 +1,6 @@
+JULIA_PREFIX = abspath(joinpath(Base.source_path(), "..", "..", ".."))
+
+if !("JULIA_PKGDIR" in keys(ENV))
+    ENV["JULIA_PKGDIR"] = joinpath(JULIA_PREFIX, "share", "julia", "site")
+    pop!(Base.LOAD_CACHE_PATH)
```
Comment (jakirkham) [inline recipe/juliarc.jl]: Need to change this line to `Base.LOAD_CACHE_PATH[1] = ENV["JULIA_PKGDIR"]` now that we have removed the `Pkg.init()` stuff.

File: recipe/juliarc.jl [hunk 8]
```diff
@@ -0,0 +1,6 @@
+JULIA_PREFIX = abspath(joinpath(Base.source_path(), "..", "..", ".."))
+
+if !("JULIA_PKGDIR" in keys(ENV))
+    ENV["JULIA_PKGDIR"] = joinpath(JULIA_PREFIX, "share", "julia", "site")
+    Base.LOAD_CACHE_PATH[1] = ENV["JULIA_PKGDIR"]
```
Comment (jakirkham) [inline recipe/juliarc.jl]: Sorry I think my edits to that comment didn't show up.

This needs to be `Base.LOAD_CACHE_PATH[1] = joinpath(ENV["JULIA_PKGDIR"], "lib", string("v", join(split(string(VERSION), ".")[1:2], ".")))`.

This makes it `$PREFIX/share/julia/site/lib/v0.5` for 0.5.x.

File: recipe/build.sh (hunk 3 above)
Comment (jakirkham) [reply recipe/build.sh]: Ah. So what sort of stuff ends up there then?

File: recipe/juliarc.jl (hunk 1 above)
Comment (jakirkham) [reply recipe/juliarc.jl]: > won't that force them to reinstall julia?

We do the same thing with `python` as well. Though `conda` is pretty smart about this. It caches all packages. Also it will try to use hard links or soft links to cutdown on the footprint of each reinstall. So many installs of the same package end up being very cheap.

> conda doesn't manage julia packages

Not currently no.

> ...you can do what you describe in julia by setting different values for JULIA_PKGDIR.

Indeed. That was the point of the wrapper and now this `.juliarc` will take on that role.

Comment (tkelman) [reply recipe/juliarc.jl]: It won't cache julia packages. I don't think you should put user-installed julia packages under the language install tree.

Comment (jakirkham) [reply recipe/juliarc.jl]: > It won't cache julia packages.

Sorry, what won't?

> I don't think you should put user-installed julia packages under the language install tree.

Ok. Should we be using `local` or something to disambiguate?

Comment (tkelman) [reply recipe/juliarc.jl]: conda won't cache julia packages if they're placed as extra content inside a different installed package (the julia language)

most julia users would expect packages to be shared between different methods of installing the same minor version of julia. if conda starts managing julia packages (I'd advise against this until after the julia package manager has been rewritten) then you could add them to the LOAD_PATH without involving julia's package manager at all.

**Comment by jakirkham** [comment]
LGTM. This is consistent with its behavior from before and removes the wrapper script (the intended goal). While there may be other discussions/changes that need to occur, would recommend we save them for new issues/PRs.

**Comment by jakirkham** [comment]
Thanks @dfornika. Also thanks @tkelman for the review.

File: recipe/juliarc.jl (hunk 1 above)
Comment (jakirkham) [reply recipe/juliarc.jl]: > conda won't cache julia packages if they're placed as extra content inside a different installed package (the julia language)

Right, but the plan would be to package Julia packages as `conda` packages. At which point `conda` would be able to handle them like any other `conda` package.

> most julia users would expect packages to be shared between different methods of installing the same minor version of julia.

This will be possible as we add `conda` packages of julia packages.

> if conda starts managing julia packages (I'd advise against this until after the julia package manager has been rewritten) then you could add them to the LOAD_PATH without involving julia's package manager at all.

Ok noted.

I think this discussion has diverged quite a bit from the original PR, which is my fault. We should move this discussion about how to manage Julia packages within the context of conda into a new issue so that we can discuss this further. Would hate to have this discussion buried in outdated comment of an unrelated diff.

Comment (tkelman) [reply recipe/juliarc.jl]: Both ways - a non-conda install of julia should be able to see packages added by running Pkg.add within a conda-installed julia. It might not see Julia packages added via conda install, but I don't think you should go there.

Comment (jakirkham) [reply recipe/juliarc.jl]: Have open issue ( https://github.com/conda-forge/julia-feedstock/issues/14 ) to discuss how best to manage Julia packages with `conda`. TBH this is something we are going to want to hammer out sooner rather than later. So we should keep an eye open for practical near term solutions.
